File: app/src/main/java/com/song/deviceinfo/ui/app/AppFragment.java
package com.song.deviceinfo.ui.app;

import com.song.deviceinfo.ui.base.BaseAdapter;
import com.song.deviceinfo.ui.base.BaseViewModel;
import com.song.deviceinfo.ui.base.NormalAdapter;
import com.song.deviceinfo.ui.base.NormalFragment;

import java.util.List;

import androidx.core.util.Pair;
import androidx.lifecycle.ViewModelProviders;

/**
 * Created by chensongsong on 2020/9/22.
 */
public class AppFragment extends NormalFragment {

    @Override
    protected BaseAdapter crateAdapter() {
        return new NormalAdapter(getContext());
    }

    @Override
    protected BaseViewModel crateViewModel() {
        return ViewModelProviders.of(this).get(AppViewModel.class);
    }

    @Override
    protected List<Pair<String, String>> getNormalInfo() {
        return ((AppViewModel) viewModel).getAppInfo(getContext());
    }
}

File: app/src/main/java/com/song/deviceinfo/ui/app/AppFragment.kt
package com.song.deviceinfo.ui.app

import androidx.core.util.Pair
import androidx.lifecycle.ViewModelProviders
import com.song.deviceinfo.ui.base.BaseAdapter
import com.song.deviceinfo.ui.base.BaseViewModel
import com.song.deviceinfo.ui.base.NormalAdapter
import com.song.deviceinfo.ui.base.NormalFragment

/**
 * Created by chensongsong on 2020/9/22.
 */
class AppFragment : NormalFragment() {
    
    override fun createAdapter(): BaseAdapter<Pair<String, String>, *>? {
        return NormalAdapter(requireContext())
    }
    
    override fun createViewModel(): BaseViewModel<Pair<String, String>>? {
        return ViewModelProviders.of(this).get(AppViewModel::class.java)
    }
    
    override val normalInfo: List<Pair<String, String>>
        get() = (viewModel as AppViewModel).getAppInfo(requireContext())
}
